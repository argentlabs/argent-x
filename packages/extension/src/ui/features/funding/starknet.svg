<svg width="40" height="40" viewBox="0 0 40 40" fill="none" xmlns="http://www.w3.org/2000/svg">
<circle cx="20" cy="20" r="20" fill="currentColor"/>
<path d="M8 20C8 13.3726 13.3726 8 20 8C26.6274 8 32 13.3726 32 20C32 26.6274 26.6274 32 20 32C13.3726 32 8 26.6274 8 20Z" fill="black"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M24.8771 17.2021L24.61 16.3769C24.5557 16.2091 24.4232 16.0785 24.2548 16.027L23.4255 15.772C23.3108 15.7369 23.3098 15.575 23.4237 15.5381L24.2492 15.271C24.4167 15.2167 24.5473 15.0842 24.5991 14.9158L24.8538 14.0865C24.8889 13.9721 25.0508 13.9708 25.0877 14.085L25.3548 14.9102C25.4091 15.0777 25.5416 15.2083 25.71 15.2601L26.5393 15.5148C26.6541 15.5502 26.6553 15.7118 26.5412 15.7487L25.7156 16.0158C25.5481 16.0701 25.4175 16.2029 25.3657 16.3713L25.111 17.2003C25.076 17.315 24.914 17.3163 24.8771 17.2021Z" fill="currentColor"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M10.4404 19.8443C10.5679 18.7843 11.0621 17.79 11.8541 16.9437C12.6421 16.1055 13.8731 15.4593 15.239 15.3736C15.9131 15.3286 16.5975 15.4037 17.2266 15.5781C17.853 15.7518 18.4399 16.0049 18.9689 16.3031C19.2332 16.4529 19.4804 16.6167 19.7268 16.782L20.3832 17.279L21.4079 18.0875C22.073 18.6171 22.7077 19.1069 23.2944 19.5114C23.8866 19.9166 24.391 20.2125 24.8728 20.386C25.3502 20.5752 25.9533 20.6236 26.7506 20.4232C27.1464 20.332 27.5593 20.1654 27.9999 19.9848C28.2485 19.8799 28.5025 19.7635 28.7666 19.6424C28.9679 19.5502 29.175 19.4552 29.39 19.3607C29.3619 19.6784 29.3147 19.9901 29.2495 20.2965C29.1033 20.9837 28.8667 21.6447 28.5518 22.2898C28.0852 23.2115 27.4362 24.1105 26.49 24.8386C25.5618 25.5726 24.2871 26.0842 22.9894 26.1611C21.6911 26.2489 20.4561 25.9629 19.4054 25.5326C18.3506 25.0961 17.4391 24.5225 16.6313 23.8893C16.4083 23.7143 16.2885 23.6151 16.1238 23.4758L15.6637 23.0728C15.508 22.9509 15.3547 22.817 15.2017 22.6834C15.0503 22.5512 14.8044 22.34 14.6518 22.2204C14.0718 21.7353 13.4947 21.2527 12.8652 20.8314C12.1508 20.3495 11.3233 19.8745 10.4404 19.8443ZM22.8507 24.1568C23.4827 24.2908 24.1766 24.3342 24.8678 24.2474C25.5162 24.167 26.1493 23.9505 26.6942 23.6522C26.4931 23.8608 26.2733 24.059 26.0326 24.2442C25.2228 24.8846 24.0867 25.3447 22.945 25.4124C21.7876 25.4907 20.6625 25.2367 19.6909 24.8391C18.7113 24.4335 17.8585 23.8982 17.094 23.299C16.8819 23.1326 16.7695 23.0395 16.613 22.9073L16.126 22.4822C15.9896 22.3754 15.8523 22.2558 15.6974 22.1205C15.5533 21.9947 15.384 21.8468 15.2189 21.7169L15.1251 21.6385C14.5523 21.1593 13.9471 20.6531 13.2835 20.2089C12.9285 19.9695 12.55 19.7391 12.1148 19.5329L12.1076 19.5295C11.9168 19.4416 11.739 19.3657 11.573 19.3166C12.0858 19.1082 12.5761 18.9552 13.0401 18.8689C13.839 18.7194 14.4901 18.8103 14.9989 19.0315C15.5114 19.2428 15.9767 19.5815 16.4781 20.0249C16.6025 20.1353 16.7281 20.256 16.8525 20.376L17.2651 20.7902C17.6144 21.1373 17.9833 21.4792 18.3767 21.8065C19.16 22.462 20.046 23.0812 21.1051 23.5739C21.6365 23.8164 22.2154 24.0265 22.8507 24.1568Z" fill="currentColor"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M14.6387 25.3107C14.6387 25.6243 14.3843 25.8787 14.0706 25.8787C13.757 25.8787 13.5029 25.6243 13.5029 25.3107C13.5029 24.9971 13.757 24.7427 14.0706 24.7427C14.3843 24.7427 14.6387 24.9971 14.6387 25.3107Z" fill="currentColor"/>
</svg>
