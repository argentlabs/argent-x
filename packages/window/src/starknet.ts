import type {
  AccountChangeEventHandler,
  AddStarknetChainParameters,
  ConnectedStarknetWindowObject,
  NetworkChangeEventHandler,
  StarknetWindowObject,
  SwitchStarknetChainParameter,
  WatchAssetParameters,
} from "get-starknet-core"
import type { ProviderInterface } from "starknet"

import { MessageAccount } from "./account"
import { userEventHandlers } from "./eventHandlers"
import { BidirectionalExchange } from "./messages/exchange/bidirectional"
import { StarknetMethods } from "./types"

export type Variant = "argentX" | "argentWebWallet"

export interface GetArgentStarknetWindowObject {
  id: Variant
  icon: string
  name: string
  version: string
  host: string
}

function updateStarknetWindowObject(
  windowObject: StarknetWindowObject,
  provider: ProviderInterface,
  remoteHandle: BidirectionalExchange<StarknetMethods, {}>,
  walletAddress: string,
): ConnectedStarknetWindowObject {
  if (windowObject.isConnected) {
    return windowObject
  }

  const valuesToAssign: Pick<
    ConnectedStarknetWindowObject,
    "isConnected" | "chainId" | "selectedAddress" | "account" | "provider"
  > = {
    isConnected: true,
    chainId: provider.chainId,
    selectedAddress: walletAddress,
    account: new MessageAccount(provider, walletAddress, remoteHandle),
    provider,
  }

  return Object.assign(windowObject, valuesToAssign)
}

export const getArgentStarknetWindowObject = (
  options: GetArgentStarknetWindowObject,
  provider: ProviderInterface,
  remoteHandle: BidirectionalExchange<StarknetMethods, {}>,
): StarknetWindowObject => {
  const wallet: StarknetWindowObject = {
    ...options,
    isConnected: false,
    provider,
    async request(call) {
      switch (call.type) {
        case "wallet_addStarknetChain": {
          const params = call.params as AddStarknetChainParameters
          return remoteHandle.call("addStarknetChain", params)
        }
        case "wallet_switchStarknetChain": {
          const params = call.params as SwitchStarknetChainParameter
          return remoteHandle.call("switchStarknetChain", params)
        }
        case "wallet_watchAsset": {
          const params = call.params as WatchAssetParameters
          return remoteHandle.call("watchAsset", params)
        }
        default:
          throw new Error("not implemented")
      }
    },
    async enable(ops) {
      if (ops?.starknetVersion === "v3") {
        throw Error("not implemented")
      }
      const [selectedAddress] = await remoteHandle.call("enable")

      updateStarknetWindowObject(
        wallet,
        provider,
        remoteHandle,
        selectedAddress,
      )

      return [selectedAddress]
    },
    async isPreauthorized() {
      const loginStatus = await remoteHandle.call("getLoginStatus")
      return loginStatus.isLoggedIn && loginStatus.isPreauthorized
    },
    on: (event, handleEvent) => {
      if (event === "accountsChanged") {
        userEventHandlers.push({
          type: event,
          handler: handleEvent as AccountChangeEventHandler,
        })
      } else if (event === "networkChanged") {
        userEventHandlers.push({
          type: event,
          handler: handleEvent as NetworkChangeEventHandler,
        })
      } else {
        throw new Error(`Unknwown event: ${event}`)
      }
    },
    off: (event, handleEvent) => {
      if (event !== "accountsChanged" && event !== "networkChanged") {
        throw new Error(`Unknwown event: ${event}`)
      }

      const eventIndex = userEventHandlers.findIndex(
        (userEvent) =>
          userEvent.type === event && userEvent.handler === handleEvent,
      )

      if (eventIndex >= 0) {
        userEventHandlers.splice(eventIndex, 1)
      }
    },
  }

  // TODO: handle network and account changes

  return wallet
}
