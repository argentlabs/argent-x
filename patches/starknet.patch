diff --git a/dist/index.mjs b/dist/index.mjs
index f23eb09368c9c26dde230d8389cda263db4981a2..d89f258fc74849693da88244b936e53900fe9775 100644
--- a/dist/index.mjs
+++ b/dist/index.mjs
@@ -3376,11 +3376,7 @@ function validateAndParseEthAddress(address) {
 }
 
 // src/utils/fetchPonyfill.ts
-import makeFetchCookie from "fetch-cookie";
-import isomorphicFetch from "isomorphic-fetch";
-var fetchPonyfill_default = IS_BROWSER && window.fetch.bind(window) || // use built-in fetch in browser if available
-!isUndefined(global) && makeFetchCookie(global.fetch) || // use built-in fetch in node, react-native and service worker if available
-isomorphicFetch;
+var fetchPonyfill_default = globalThis.fetch.bind(globalThis);
 
 // src/utils/provider.ts
 var provider_exports = {};
