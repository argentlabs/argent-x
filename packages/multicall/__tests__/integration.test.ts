import { SequencerProvider } from "starknet"
import { beforeAll, describe, expect, test } from "vitest"

import { Multicall } from ".."

describe("Multicall", () => {
  let mc: Multicall
  beforeAll(() => {
    mc = new Multicall(
      new SequencerProvider({
        baseUrl: "https://alpha4-2.starknet.io/",
      }),
      undefined,
      {
        batchInterval: 10, // 10ms
      },
    )
  })

  test("should aggregate multiple calls into one multicall", async () => {
    const results = await Promise.all([
      // promises resolve at the same time as they use multicall contract
      mc.call({
        contractAddress:
          "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
        entrypoint: "balanceOf",
        calldata: [
          "0x04a79cA7FDE3dd9C5CBadcBDCB39f95A0619da26767af0B52fD0901cd556a035".toLowerCase(),
        ],
      }),
      mc.call({
        contractAddress:
          "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
        entrypoint: "balanceOf",
        calldata: [
          "0x0472eb42746E4b7b426B6DC45B9ac0345bA38502d0928209016F8a1323330CF4".toLowerCase(),
        ],
      }),
    ])

    expect(results).toMatchSnapshot()
  })
  test("should partially error with a single error", async () => {
    const results = await Promise.allSettled([
      // promises resolve at the same time as they use multicall contract
      mc.call({
        contractAddress:
          "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
        entrypoint: "balanceOf",
        calldata: [
          "0x0472eb42746E4b7b426B6DC45B9ac0345bA38502d0928209016F8a1323330CF4".toLowerCase(),
        ],
      }),
      mc.call({
        contractAddress:
          "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
        entrypoint: "get_balance", // this will fail
        calldata: [
          "0x04a79cA7FDE3dd9C5CBadcBDCB39f95A0619da26767af0B52fD0901cd556a035".toLowerCase(),
        ],
      }),
    ])

    expect(results.map((x) => x.status)).toMatchInlineSnapshot(`
      [
        "fulfilled",
        "rejected",
      ]
    `)
  })
  test("should partially error with multiple errors", async () => {
    const results = await Promise.allSettled([
      // promises resolve at the same time as they use multicall contract
      mc.call({
        contractAddress:
          "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
        entrypoint: "get_balance", // this will fail
        calldata: [
          "0x0472eb42746E4b7b426B6DC45B9ac0345bA38502d0928209016F8a1323330CF4".toLowerCase(),
        ],
      }),
      mc.call({
        contractAddress:
          "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
        entrypoint: "balanceOf",
        calldata: [
          "0x0472eb42746E4b7b426B6DC45B9ac0345bA38502d0928209016F8a1323330CF4".toLowerCase(),
        ],
      }),
      mc.call({
        contractAddress:
          "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
        entrypoint: "get_balance", // this will fail
        calldata: [
          "0x04a79cA7FDE3dd9C5CBadcBDCB39f95A0619da26767af0B52fD0901cd556a035".toLowerCase(),
        ],
      }),
    ])

    expect(results.map((x) => x.status)).toMatchInlineSnapshot(`
      [
        "rejected",
        "fulfilled",
        "rejected",
      ]
    `)
  })
  test("should error when all fail", async () => {
    const results = await Promise.allSettled([
      // promises resolve at the same time as they use multicall contract
      mc.call({
        contractAddress:
          "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
        entrypoint: "get_balance", // this will fail
        calldata: [
          "0x0472eb42746E4b7b426B6DC45B9ac0345bA38502d0928209016F8a1323330CF4".toLowerCase(),
        ],
      }),
      mc.call({
        contractAddress:
          "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
        entrypoint: "get_balance", // this will fail
        calldata: [
          "0x04a79cA7FDE3dd9C5CBadcBDCB39f95A0619da26767af0B52fD0901cd556a035".toLowerCase(),
        ],
      }),
    ])

    expect(results.map((x) => x.status)).toMatchInlineSnapshot(`
      [
        "rejected",
        "rejected",
      ]
    `)
  })
})
