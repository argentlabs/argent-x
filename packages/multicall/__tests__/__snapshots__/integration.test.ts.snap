// Vitest Snapshot v1

exports[`Multicall > should aggregate multiple calls into one multicall 1`] = `
[
  [
    "0x14aa3c31ab2a7e16",
    "0x0",
  ],
  [
    "0x2386f26fc10000",
    "0x0",
  ],
]
`;
